rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,126 @@
+
+for target in alice_mut.targets<Eats>() {
+	// target == apple
+	// target == banana
+}
+
+let query = world.query<(Entity, Targets<Eats>)>();
+for entity, targets in query.iter(&world) {
+	// entity == alice
+	for target in targets {
+		// target == apple
+		// target == banana
+	}
+}
+
+let query = world.query<Entity, With<Targets<Eats>>>();
+for entity in query.iter(&world) {
+	// entity == alice
+}
+
+```
+
+Dynamic queries would also be able to match on specific pairs:
+```rust
+let query = QueryBuilder::new::<Entity>(&mut World)
+	.with_pair::<Eats>(apple)
+	.build();
+
+query.single(&world); // == alice
+```
+
+## Implementation strategy
+
+There are a number of technical blockers before even a minimal version of relationships could be implemented in bevy. I've broken them down below:
+#### Archetype Cleanup
+In order to fit each pair into a single id we lose the entity generation (see this [article](https://ajmmertens.medium.com/doing-a-lot-with-a-little-ecs-identifiers-25a72bd2647) for more details). As a consequence when an entity is despawned we need to ensure that all the archetypes that included that entity as a target are also destroyed so that a future entity re-using that id is not misinterpreted as being the old target. This is non-trivial.
+
+`bevy_ecs` currently operates under the assumption that archetype and component ids are dense and strictly increasing. In order to break that invariant we need to address several issues:
+##### Query and System Caches
+Query's caches of matching tables and archetypes are updated each time the query is accessed by iterating through the archetypes that have been created since it was last accessed. As the number of archetypes in bevy is expected to stabilize over the runtime of a program this isn't currently an issue. However with the increased archetype fragmentation caused by fragmenting relationships and the new wrinkle of archetype deletion the updating of the query caches is poised to become a performance bottleneck.
+
+Luckily observers [#10839](https://github.com/bevyengine/bevy/pull/10839) can help solve this problem. By sending ECS events each time an archetype is created and creating an observer for each query we can target the updates to query caches so that we no longer need to iterate every new archetype. Similarly we can fire archetype deletion events to cause queries to remove the archetype from their cache.
+
+In order to implement such a mechanism we need some way for the observer to have access to the query state in order to update it. This can be done by moving the query state into a component on the associated entity (WIP branch here: [queries-as-entities](https://github.com/james-j-obrien/bevy/tree/queries-as-entities)) and will require further refactors to the cache to optimize for removal. However this is largely blocked by the deletion of entities in the render world tracked here: [#12144](https://github.com/bevyengine/bevy/issues/12144).
```

Comment:
How are we to handle all of the potential interactions that an entity supports when dealing with soundness?

 * A `Query` as a component being `std::mem::swap`'ed with another with a different component access or different `WorldId`. This would trivially enable unsound behavior without use of `unsafe`.
 * `Query` state components being queried by other queries, and then used to access entities that are not in the original component access set.
 * A `Query` has been despawned.

These need to be addressed in some way in this PR to ensure we're not introducing trivial unsoundness.

Reply:
I haven't spent time here expanding the details of queries as entities here, but part of that process would be taking `QueryState` private so that you can't run into those first two problem as a user.

Despawning a query is always going to be problematic but most users won't be querying for queries so unless they do a naive "despawn all entities" operation it's unlikely a user will run into this. If they do despawn a query that's used in a system it will likely result in a panic.

Reply:
> If they do despawn a query that's used in a system 

Simple solution for this is to also associate the system with the same entity (which is what I do), so that when you delete the query entity you're guaranteed to also delete the system.

Reply:
We have multiple queries per system but we can use relationships to associate them eventually. 

Users may encounter some footguns with despawning a query by accident and having arbitrary systems stop running, but that's a debate to be had when we get there, in the mean time I think a panic is fine.

Reply:
> I haven't spent time here expanding the details of queries as entities here, but part of that process would be taking `QueryState` private so that you can't run into those first two problems as a user.

This probably deserves mentioning. Any potential soundness hole in the implementation is a hard blocker here.

Another potential item that might need fully exploring in this RFC is how queries as entities integrates with existing system accesses. An `Access<ComponentId>` will always conflict on the query component unless something is changed in the access model, or the conflict is explicitly ignored. `Access<ArchetypeComponentId>` may also need to factor this in as well.

Reply:
Since `QueryState` doesn't need to be mutated to iterate the query systems can all have read access to the `QueryState` component without causing issues for soundness or parallelism.

Reply:
That's true within a system, but the schedule executor needs to mutate them to update their archetypes before the  systems that use them can run, potentially while other systems are running. Trivial in single threaded executors, but a way to prevent aliasing in the multithreaded executor is needed.

Reply:
As part of queries as entities we would also be transitioning to relying on observers to update the query caches, so those updates would be applied during command application.

Is there some other mutation that needs to be done during scheduling?